### PR TITLE
Avoid sharing handles between threads

### LINKED ISSUE #1109: Avoid sharing handles for the same device between threads
Currently, when the user code is multi-threaded AND each thread uses the same GPU device, the same handle is used between threads for library calls. However, currently we are calling library functions using `with nogil`.

https://github.com/cupy/cupy/blob/v4.0.0rc1/cupy/cuda/device.pyx#L22-L25
https://github.com/cupy/cupy/blob/v4.0.0rc1/cupy/random/generator.py#L552-L583

This means that there may be two issues:

1. Library function invocations using the same handle may run concurrently.
2. Stream set via `setStream` call (prior to the library function invocation) may be overwritten by the subsequent call.

Here is a quote of Thread Safetiness section from API reference of each library.

## cuBLAS

> The library is thread safe and its functions can be called from multiple host threads, even with the same handle. When multiple threads share the same handle, extreme care needs to be taken when the handle configuration is changed because that change will affect potentially subsequent CUBLAS calls in all threads. It is even more true for the destruction of the handle. So it is not recommended that multiple thread share the same CUBLAS handle.

Both issue 1. and 2.
(we're using `cublasSetMathMode` which changes the configuration for the handle)

## cuDNN

> For multithreaded applications that use the same device from different threads, the recommended programming model is to create one (or a few, as is convenient) cuDNN handle(s) per thread and use that cuDNN handle for the entire life of the thread.

> The library is thread safe and its functions can be called from multiple host threads, as long as threads to do not share the same cuDNN handle simultaneously.

Both issue 1. and 2.

## cuFFT (handle = plan)

> cuFFT APIs are thread safe as long as different host threads execute FFTs using different plans and the output data are disjoint.

No issue in CuPy, as we can expect that `Plan1d` (our private API that creates plan) instances are not shared between threads.

## cuSOLVER

> The library is thread safe and its functions can be called from multiple host threads.

Unclear.
Maybe both issue 1 and 2?

## cuSPARSE

> The library is thread safe and its functions can be called from multiple host threads.

Unclear.
No issue in CuPy, as our cuSPARSE invocation currently does not use `with nogil`.

## cuRAND (handle = generator)

Unclear; no description on API reference.

### PR DESCRIPTION
**Pull request by kmaehashi**
Fixes #1109 and #2045.

### CODE REVIEW

**Comment by kmaehashi** [comment]
@okuta Could you take a look if this fix is acceptable? I'd like to do the same thing for all handles.

**Comment by okuta** [comment]
OK, It is acceptable.

File: cupy/cuda/device.pyx [hunk 1]
```diff
@@ -189,11 +206,16 @@ cdef class Device:
...
             raise RuntimeError(
                 'Current cupy only supports cusolver in CUDA 8.0')
-        if self.id in _cusolver_handles:
-            return _cusolver_handles[self.id]
+
+        _cusolver_handles = getattr(_thread_local, '_cusolver_handles', None)
+        if _cusolver_handles is None:
+            _cusolver_handles = {}
+            setattr(_thread_local, '_cusolver_handles', _cusolver_handles)
+        elif self.id in _cusolver_handles:
```
Comment (okuta) [inline cupy/cuda/device.pyx]:
```suggestion
        handle = _cusolver_handles.get(self.id, None)
        if handle is not None:
            return handle.handle
```

File: cupy/cuda/device.pyx [hunk 2]
```diff
@@ -17,17 +18,37 @@ except ImportError:
...
 cdef dict _cusolver_sp_handles = {}
 cdef dict _cusparse_handles = {}
 cdef dict _compute_capabilities = {}
 
 
+cpdef int get_device_id() except? -1:
+    return runtime.getDevice()
+
+
+cpdef Device get_device():
```
Comment (okuta) [inline cupy/cuda/device.pyx]:
```suggestion
cpdef Device _get_device():
```

File: cupy/cuda/device.pyx [hunk 3]
```diff
@@ -37,11 +58,7 @@ cpdef size_t get_cublas_handle() except? 0:
 
 
 cpdef size_t get_cusolver_handle() except? 0:
-    dev_id = get_device_id()
-    ret = _cusolver_handles.get(dev_id, None)
-    if ret is not None:
-        return ret
-    return Device().cusolver_handle
+    return get_device().cusolver_handle
```
Comment (okuta) [inline cupy/cuda/device.pyx]:
```suggestion
    return _get_device().cusolver_handle
```

**Comment by pentschev** [comment]
@kmaehashi thanks for starting with this issue, don't hesitate to ask me if you need some extra hands to help with it.

**Comment by kmaehashi** [comment]
Ready for review.

**Comment by pentschev** [comment]
FWIW, I just tested this branch with my multithreaded use cases and it works great. Thanks a lot for this PR, @kmaehashi!

**Comment by kmaehashi** [comment]
@pentschev Thanks for testing PR! We will proceed to review.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
```
E   AssertionError: Parameterized test failed.
E   
E   Base test method: TestLschol.test_csrmatrix
E   Test parameters:
E     dtype: <type 'numpy.float32'>
E   
E   AssertionError: 
E   Fail: 10, Success: 0
E   
E   The first error message:
E   Traceback (most recent call last):
E     File "cupy/testing/condition.py", line 61, in <lambda>
E       lambda: f(ins, *args[1:], **kwargs),
E     File "/work/cupy/tests/cupyx_tests/linalg_tests/sparse_tests/test_solve.py", line 61, in test_csrmatrix
E       x = cupyx.linalg.sparse.lschol(A, b)
E     File "cupyx/linalg/sparse/solve.py", line 49, in lschol
E       handle = device.get_cusolver_sp_handle()
E     File "cupy/cuda/device.pyx", line 57, in cupy.cuda.device.get_cusolver_sp_handle
E       cpdef get_cusolver_sp_handle():
E     File "cupy/cuda/device.pyx", line 58, in cupy.cuda.device.get_cusolver_sp_handle
E       return _get_device().cusolver_sp_handle
E     File "cupy/cuda/device.pyx", line 213, in cupy.cuda.device.Device.cusolver_sp_handle.__get__
E       'cusolver_sp_handles', cusolver.spCreate, cusolver.spDestroy)
E   AttributeError: 'module' object has no attribute 'spDestroy'```

**Comment by jakirkham** [comment]
Looks like this is just a matter of adding a `spDestroy` function that wraps [`cusolverSpDestroy`]( https://docs.nvidia.com/cuda/cusolver/index.html#cusparsedestroy ). Does that sound right?

**Comment by kmaehashi** [comment]
Thanks, locally confirmed `tests/cupyx_tests/linalg_tests/sparse_tests/test_solve.py` pass.

@jakirkham Yes, I added it.

**Comment by okuta** [comment]
travis ci error.
```
    cupy/cuda/cusolver.cpp:2370:80: error: ‘cusolverSpDestroy’ was not declared in this scope
             __pyx_v_status = cusolverSpDestroy(((cusolverSpHandle_t)__pyx_v_handle));```

**Comment by kmaehashi** [comment]
Test failures are not related to this PR.

**Comment by jakirkham** [comment]
Is there anything else needed before integrating this PR? Anything that either of you need help with?

**Comment by okuta** [comment]
LGTM!

**Comment by okuta** [comment]
I will backport this PR after #2121 merged.

**Comment by jakirkham** [comment]
Thank you! 😊

**Comment by pentschev** [comment]
Thank you all for the effort here, this fix is very much welcome for the work I've been doing! :)

**Pull request merged**